File: go/lib/decrypt.go
package lib

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/ecdh"
	"crypto/ecdsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/pem"
	"fmt"
	"os"
)

type EncryptedPayload struct {
	Algorithm string `json:"alg"`
	Data      string `json:"data"`
	Env       string `json:"env"`
	Key       string `json:"key"`
}

func DecryptData(privateKeyPath string, payload EncryptedPayload) ([]byte, error) {
	// 1. Load private key
	privateKeyBytes, err := os.ReadFile(privateKeyPath)
	if err != nil {
		return nil, fmt.Errorf("failed to read private key: %v", err)
	}

	// Parse PEM block
	block, _ := pem.Decode(privateKeyBytes)
	if block == nil {
		return nil, fmt.Errorf("failed to parse PEM block")
	}

	// Parse private key
	privateKey, err := x509.ParseECPrivateKey(block.Bytes)
	if err != nil {
		return nil, fmt.Errorf("failed to parse private key: %v", err)
	}

	// 2. Decode ephemeral public key from base64
	publicKeyDER, err := base64.StdEncoding.DecodeString(payload.Key)
	if err != nil {
		return nil, fmt.Errorf("failed to decode public key: %v", err)
	}

	// Parse public key
	publicKey, err := x509.ParsePKIXPublicKey(publicKeyDER)
	if err != nil {
		return nil, fmt.Errorf("failed to parse public key: %v", err)
	}

	// 3. Create ECDH keys
	curve := ecdh.P384()

	privKey, err := curve.NewPrivateKey(privateKey.D.Bytes())
	if err != nil {
		return nil, fmt.Errorf("failed to create ECDH private key: %v", err)
	}

	// Concatenate X and Y coordinates with 0x04 prefix (uncompressed point format)
	pubKeyBytes := append([]byte{0x04}, append(publicKey.(*ecdsa.PublicKey).X.Bytes(), publicKey.(*ecdsa.PublicKey).Y.Bytes()...)...)
	pubKey, err := curve.NewPublicKey(pubKeyBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to create ECDH public key: %v", err)
	}

	// 4. Compute shared secret
	sharedSecret, err := privKey.ECDH(pubKey)
	if err != nil {
		return nil, fmt.Errorf("ECDH failed: %v", err)
	}

	// 5. Generate AES key by hashing shared secret
	hash := sha256.Sum256(sharedSecret)
	aesKey := hash[:]

	// 6. Decode encrypted data
	encryptedData, err := base64.StdEncoding.DecodeString(payload.Data)
	if err != nil {
		return nil, fmt.Errorf("failed to decode encrypted data: %v", err)
	}

	// 7. Create AES cipher
	aesCipher, err := aes.NewCipher(aesKey)
	if err != nil {
		return nil, fmt.Errorf("failed to create AES cipher: %v", err)
	}

	// 8. Create CBC decrypter with zero IV
	iv := make([]byte, aes.BlockSize)
	mode := cipher.NewCBCDecrypter(aesCipher, iv)

	// 9. Decrypt data
	decrypted := make([]byte, len(encryptedData))
	mode.CryptBlocks(decrypted, encryptedData)

	// 10. Remove padding if present
	if len(decrypted) > 0 {
		paddingLen := int(decrypted[len(decrypted)-1])
		if paddingLen > 0 && paddingLen <= aes.BlockSize {
			decrypted = decrypted[:len(decrypted)-paddingLen]
		}
	}

	return decrypted, nil
}

File: python/decrypt.py
from cryptography.hazmat.primitives import serialization, hashes
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
import base64

def decrypt_data(private_key_path: str, payload: dict) -> bytes:
    """
    Decrypts data that was encrypted using ECDH key exchange and AES-256-CBC encryption.

    Args:
        private_key_path (str): Path to the private key PEM file used for ECDH
        payload (dict): Dictionary containing the encrypted data from Notehub event:
            - key: Base64 encoded ephemeral public key
            - data: Base64 encoded encrypted data

    Returns:
        bytes: The decrypted data with padding removed

    Raises:
        Various cryptography exceptions if decryption fails
    """
    # 0. Check if payload is valid
    if not payload or not isinstance(payload, dict):
        raise ValueError("Invalid payload")
    
    # 1. Check if algorithm is supported
    if payload['alg'] != 'secp384r1-aes256cbc':
        raise ValueError("Unsupported algorithm")
    
    # 3. Decode the base64 ephemeral public key
    public_key_bytes = base64.b64decode(payload['key'])
    peer_public_key = serialization.load_der_public_key(public_key_bytes)

    # 4. Load private key and derive shared secret
    with open(private_key_path, 'rb') as f:
        private_key = serialization.load_pem_private_key(
            f.read(),
            password=None
        )
    shared_secret = private_key.exchange(ec.ECDH(), peer_public_key)
    print(f"Shared secret: {shared_secret.hex()}")

    # 5. Hash the shared secret to get AES key
    digest = hashes.Hash(hashes.SHA256())
    digest.update(shared_secret)
    aes_key = digest.finalize()
    print(f"AES key: {aes_key.hex()}")

    # 6. Decode the encrypted data
    encrypted_data = base64.b64decode(payload['data'])
    print(f"Encrypted data: {encrypted_data.hex()}")

    # 7. Decrypt with zero IV
    cipher = Cipher(algorithms.AES256(aes_key), modes.CBC(b'\x00' * 16))
    decryptor = cipher.decryptor()
    decrypted = decryptor.update(encrypted_data) + decryptor.finalize()
    
    # 8. Remove padding if present
    try:
        padding_length = decrypted[-1]
        if padding_length <= 16:
            decrypted = decrypted[:-padding_length]
    except:
        pass

    return decrypted